### ISSUE TITLE
[Docs]: Can I precache routes? (Next)

### ISSUE THREAD

**Issue by timohausmann** [serwist/serwist#49]
### Provide environment information

Hey, I am new to Service Workers. I try to make my next 14 pwa work fully offline.
Is there a way to prefetch all routes? As far as my testing goes, serwist out of the box only caches already requested routes. 

### Describe the problem

When I set `Network > Throttling > Offline` and use a `<Link>` to an unvisited route, I still leave my app and get hit with a hard Chrome 404. Is this normal? 

### Expected behavior

I would like to define routes that should be prefetched. If I hit a 404 I would expect to stay in my app and show a message instead of "crashing" it. 

### Screenshots (if relevant)

![Bildschirmfoto 2024-01-22 um 16 15 22](https://github.com/serwist/serwist/assets/2311113/da941907-3dd8-416a-8e53-9ca4f1a45ff9)

**Comment by DuCanhGH** [comment]
You can use `additionalPrecacheEntries`.
```js
const withSerwist = withSerwistInit({
  cacheOnFrontEndNav: true,
  swSrc: "app/sw.ts",
  swDest: "public/sw.js",
  additionalPrecacheEntries: [
    { url: "/foo", revision: getRevision("./app/foo") },
    { url: "/bar", revision: getRevision("./app/bar") },
  ],
});
```
Note that you have to implement this `getRevision` yourself.

**Comment by timohausmann** [comment]
@DuCanhGH 
Thank you so much, I did not find this option in the docs! 
I will test it and report back.

**Comment by DuCanhGH** [comment]
@timohausmann https://serwist.pages.dev/docs/build/configuring/base-options/additional-precache-entries
It handles all routes indiscriminately. I wouldn't recommend precaching dynamic routes to be honest, but you may try.

**Comment by timohausmann** [comment]
Thanks for the link! E.g. `/item/:id` I only have 100 items and most of the data is in memory already anyway, I don't like defining 100 routes at build. I just don't get how to tell the browser it's save to go to to e.g. `/item/1234`. 
Maybe I can simply change the routes to `/item/?id=1234` and only precache `/item`?

**Comment by DuCanhGH** [comment]
@timohausmann yes, you can. Doesn't change much if you use searchParams in RSC though. You have to use `useSearchParams`.

**Comment by timohausmann** [comment]
Oh I’m still on pages router, no `useSearchParams` for me. 

By default when built, Next seems to preload every `<Link>` in the viewport which is picked up by the SW, clicking these works offline, cool combo so far.

I tried `additionalPrecacheEntries: ['/settings']` but that’s just the HTML response, I rather need the `/_next/...js(on)` build files for that sweet frontend routing. 

For basic routes, I discovered Next’s `router.prefetch('/settings')`, which is smart enough to fetch the related  build files, so that works already, omg!

Now I still struggle with dynamic routes. I think URL segment or GET parameter make no difference. I could loop over the IDs and prefetch all, but yikes, especially since they all use the same pageProps and html skeleton anyway (they get populated client-side based on the dynamic GET id).

Any ideas? Is there some kind of wildcard I can leverage, that returns the same js(on) for all items? 
(No idea how to get the hashed full paths like `/_next/data/igvgOxc9KeS8C_wWohMnw/en/item.json?item=RMDwtxABkVcHIgUqnK87` though)

**Comment by DuCanhGH** [comment]
@timohausmann use `router.query` and simply prefetch "/item".

**Comment by timohausmann** [comment]
It's not working because it is requesting `item.json?item=RMDwtxABkVcHIgUqnK87` specifically.
_(Failed to load response data: No data found for resource with given identifier)_

**Comment by DuCanhGH** [comment]
@timohausmann meh. If it somehow gets that search query with `router.prefetch` then I don't have anything to tell you either. If anything though, `additionalPrecacheEntries` should just work, since the json files should be precached automatically by Serwist by default (see `self.__SW_MANIFEST`).

**Comment by timohausmann** [comment]
`additionalPrecacheEntries` helped me with a static svg file but I couldn't get it working for any Next route.

This sledgehammer method is good enough for a prototype but I have to find another way:
```js
items.forEach((item) => {
    router.prefetch(`/item?item=${item.id}`);
});
```

I will update here if I find another solution. I will also try to publish a minimal repo for reproduction.

**Comment by timohausmann** [comment]
Okay, I found a way to make a single `router.prefetch('/item')` work! 

I removed all `getStaticProps` from my pages (they were redundantly fetching i18n data). I moved that to `getInitialProps` in `_app.tsx`. That makes me loose SSG, but it seems like Next stops assuming every item has unique data. All this may be easier with the app router.

My final frontier was being able to hit refresh, after adding `['/', '/de', '/fr']` to `additionalPrecacheEntries` I can seemingly refresh the home views now. Refreshing deep URLs like `/item/1234` crashes hard, sadly. Would be cool to have an emergency fallback to the root URL.

I don’t really get why .svg, .png etc. from the public folder won’t get picked up by `runtimeCaching` but adding them to `additionalPrecacheEntries` seems to work. This is all black magic to me, haha. So happy I got 95% working this week.

**Comment by DuCanhGH** [comment]
@timohausmann I, frankly, don't understand why you need `getInitialProps`, but cool I guess. That `getStaticProps` doesn't work nicely with Serwist doesn't sit nice with me, but since I haven't used Pages Router for so long, I am actually pretty clueless as well. 

> I don’t really get why .svg, .png etc. from the public folder won’t get picked up by runtimeCaching but adding them to additionalPrecacheEntries seems to work. This is all black magic to me, haha. So happy I got 95% working this week.

That's weird.  The default `runtimeCaching` array does handle these files though. Not sure why that would be the case.

**Comment by timohausmann** [comment]
I have a pretty basic pages router setup, `getInitialProps` pulls in the i18n json files in the current locale. So now it is only loaded once instead of per page (`getStaticProps`) which is easier to cache. I found my ways now, but I also feel like there is a mistake somewhere. I will try to find the time for a clean next project with app router to learn more about Serwist and investigate my issues.

**Comment by xdmx** [comment]
I've been playing a bit with this, I have a small nextjs app that I want to make it available offline and I've added:
```js
      additionalPrecacheEntries: [
        { url: '/images/background.png', revision: getRevision('./public/images/background.png') },
        { url: '/page1', revision: crypto.randomUUID() },
        { url: '/page2', revision: crypto.randomUUID() }
      ]
```

the `getRevision` simply calculates the md5 of the file, that way if it changes it updates it, while for the routes I've set a random uuid to always refetch them when the manifest gets regenerated at the next deploy

My question is: when should I expect the manifest to be refetched by the browser and see that it changed and thus refetch those resources? I made a few deploys so far but it's still using the old pages/files, as if it never refetched the manifest.

I've read in some places that it should happen automatically but it hasn't (it's been a few days since those deploys).

I've also tried to add a `useEffect` with `navigator.serviceWorker?.getRegistration()?.then(registration => registration.update())` in the main `App` in order to try to update the manifest at each refresh (the browser should prevent that and only do that when the last fetch happened more than 24h ago) and reinstalled the pwa, but still nothing changed even with that (firefox on android)

What am I missing? @DuCanhGH thoughts?
